python/modem-firmware-1.3+/nrf_cloud_provision.py

```diff
@@ -0,0 +1,350 @@
+    print("{}: {} - {}".format(custom_text, api_result.status_code, api_result.reason))
+    if print_response_txt:
+        print("Response: {}".format(api_result.text))
+
+def get_provisioning_results(api_key, bulk_ops_req_id):
+
+    print("Fetching results for {}: {}".format(BULK_OP_REQ_ID, bulk_ops_req_id))
+
+    while True:
+        print("Waiting 5s...")
+        time.sleep(5)
+
+        api_result = get_bulk_ops_result(api_key, bulk_ops_req_id)
+
+        if api_result.status_code != 200:
+            print("Failed to fetch provisioning result")
+            return None
+
+        api_result_json = api_result.json()
+
+        print("Provisioning status: " + api_result_json["status"])
+
+        if api_result_json["status"] == "IN_PROGRESS" or api_result_json["status"] == "PENDING":
+            continue
+
+        return api_result
+
+
+def parse_err_msg(err_str):
+
+    # Until the error msg is fixed by the cloud, we have to do some extra parsing...
+    # See IRIS-3758
+
+    # Search for the end of the first item, which is just a general error msg
+    err_begin_idx = err_str.find(ERR_FIND_FIRST_STR)
+
+    # Find the end of the detailed error list, which has escaped quotes
+    err_end_idx = err_str.rfind(ERR_FIND_END_STR)
+
+    if err_begin_idx == -1 or err_end_idx == -1:
+        print("Unhandled error response format")
+        return
+
+    # Inspect the first item for total number of reported errors
+    err_begin_str = err_str[:err_begin_idx]
+    err_cnt = 0
+    for s in err_begin_str.split():
+        if s.isdigit():
+            err_cnt = int(s)
+            break
+
+    if err_cnt == 0:
+        print("Warning: no errors reported")
+
+    # Get the start of the detailed error list, which is after the first item
+    err_json_str = err_str[ (err_begin_idx + len(ERR_FIND_FIRST_STR)) : err_end_idx]
+
+    # Fix the escaped quotes
+    err_json_str = literal_eval("'%s'" % err_json_str)
+
+    # Return the error count and the json formatted error dict
+    return err_cnt, json.loads(err_json_str)
+
+def update_device_list_err(dev_list, err_dict):
+
+    list_sz = len(dev_list)
+
+    # Loop through the dictionary of errors
+    for err_item in err_dict:
+
+        # The key is the error text and the value is a json array (python list)
+        # of indicies into the provisioning CSV file, and also the device list
+        idx_list = err_dict[err_item]
+
+        # Use the indicies to access the device list
+        for dev_idx in idx_list:
+            i = dev_idx -1
+            if i >= list_sz:
+                print("Reported device index out of range: {}".format(dev_idx))
+                continue
+
+            # Add the error message to the device list
+            if len(err_item) == 0 or err_item == ' ':
+                # In case of an empty error message...
+                dev_list[i][DEV_LIST_RES_IDX] = 'ERROR_UNKNOWN'
+            else:
+                dev_list[i][DEV_LIST_RES_IDX] = err_item
```

Comment:
What happens in the case that one or more devices succeed, but one or more fail too?  Does the `dev_list[i][DEV_LIST_RES_IDX]` get set to 'OK' for the successful devices in that case?

Reply:
yes, if the device isn't in the error report, it is successful and will be marked OK below.